### PR TITLE
feat: Add OpenRouter integration for unified access to 200+ AI models

### PR DESCRIPTION
**Pull request by Seyamalam**
### OpenRouter Integration

This PR adds OpenRouter integration to the GenAI Processors library, enabling unified access to 200+ AI models from multiple providers through a single API.

### Overview

OpenRouter (https://openrouter.ai) provides access to models from:
- **OpenAI**: GPT-4o, GPT-4.1
- **Anthropic**: Claude 3.7 Sonnet, Claude 4 Opus
- **Google**: Gemini Pro, Gemini Flash, Gemma models
- **Meta**: Llama 3.3 (8B, 70B, 405B), Llama 4
- **Many others**: Mistral, Cohere, and 200+ additional models

### Key Features

- **Full streaming support** with Server-Sent Events (SSE) parsing
- **Function calling** support for tool-enabled models
- **Multimodal inputs** (text + images)
- **Rich metadata** for cost tracking and usage monitoring
- **Configurable parameters** (temperature, max_tokens, etc.)
- **Robust error handling** with detailed HTTP status reporting
- **Seamless integration** with existing GenAI Processors patterns

### Files Added

- `genai_processors/contrib/openrouter_model.py` - Core processor implementation
- `genai_processors/contrib/README.md` - Documentation and usage guide  
- `genai_processors/tests/openrouter_model_test.py` - Comprehensive test suite (18 tests)
- `examples/openrouter_example.py` - Interactive demo script

### Files Modified

- `genai_processors/contrib/__init__.py` - Export new classes
- `pyproject.toml` - Add `httpx>=0.24.0` dependency for contrib

### Testing

- ✅ **18 comprehensive test cases** covering all functionality
- ✅ **Mocked HTTP responses** for reliable CI/CD
- ✅ **Integration tests** with `processor.apply_sync()`
- ✅ **Error handling** and edge case validation
- ✅ **Function calls** and multimodal content testing

### Usage Example

```python
from genai_processors.contrib import OpenRouterModel, GenerateContentConfig

# Initialize processor
processor = OpenRouterModel(
    api_key="your-openrouter-api-key",
    model_name="openai/gpt-4o",
    config=GenerateContentConfig(
        temperature=0.7,
        max_tokens=500
    )
)

# Use with streaming
async for response in processor.call("Explain quantum computing"):
    print(response.text, end='')
```

### Design Decisions

- **Contrib package placement**: OpenRouter is an external service, fitting the contrib pattern
- **SSE parsing**: Custom implementation for OpenRouter's streaming format
- **Metadata extraction**: Captures cost, usage, and model information for transparency
- **Error handling**: Comprehensive HTTP error handling with context
- **Key prefix**: Includes model name and config for proper cache isolation

### Documentation

- Comprehensive README with quickstart guide
- Inline code documentation following project standards
- Example script with multiple model demonstrations
- Clear setup instructions and environment variable configuration

### Checklist

- [x] Follows existing code patterns and style
- [x] Comprehensive test coverage (18/18 tests passing)
- [x] Documentation and examples provided
- [x] Dependencies properly declared in pyproject.toml
- [x] Error handling and edge cases covered
- [x] Integration tested with existing processor patterns

This integration significantly expands the GenAI Processors ecosystem by providing access to cutting-edge models from all major AI providers through a single, unified interface.

The implementation also follows established GenAI Processors patterns and provides seamless integration with the existing ecosystem while enabling access to the full breadth of modern AI models available through OpenRouter.

### CODE REVIEW

File: genai_processors/contrib/__init__.py [hunk 1]
```diff
@@ -12,4 +12,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
-"""Community contributed processors."""
+
+from .openrouter_model import OpenRouterModel
```
Comment (kibergus) [inline genai_processors/contrib/__init__.py:16]: Please don't import modules in __init__.py. This is contrib folder and we aim to add various integrations there. They will bring assorted dependencies with them. If we import them in __init__.py then all of them inevitably will be imported when accessing any of the contrib modules. At the very least this will slow down import and increas application footprint.

You don't need any changes in __init__.py. People can still import your module as `from genai_procesors.contrib import openrouter_model`.

File: genai_processors/contrib/openrouter_model.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+) -> dict[str, Any]:
+  """Convert ProcessorPart to OpenRouter message format."""
+  role = part.role.lower() if part.role else default_role
+
+  # Handle function calls
+  if hasattr(part.part, 'function_call') and part.part.function_call:
+    return {
+        'role': 'assistant',
+        'function_call': {
+            'name': part.part.function_call.name,
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:144]: Just `part.function_call.name`, no need for doubel part.part

File: genai_processors/contrib/openrouter_model.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+              {
+                  'type': 'image_url',
+                  'image_url': {
+                      'url': f'data:{part.mimetype};base64,{encoded_image}'
+                  },
+              }
+          ],
+      }
+
+  # Fallback to text representation
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:185]: I think it woul dbe better to fail verbosely:
```
   raise ValueError(f'Unsupported Part type: {part.mimetype}')
```

File: genai_processors/contrib/openrouter_model.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+
+  # Parse data lines
+  if line.startswith('data: '):
+    data = line[6:]
+    if data == '[DONE]':
+      return {'type': 'done'}
+
+    try:
+      return json.loads(data)
+    except json.JSONDecodeError:
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:208]: Is there a reason to suppress errors? I'd expect that receiving a malformed JSON would mean that something went wrong and we can't be sure that server would behave robustly in the future. So may be just don't handle the error and let the pipeline crash.

That is if errors are not part of regular flow in openrouter, I don't have much experience with it.

File: genai_processors/contrib/openrouter_model.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    """Key prefix for caching."""
+    return f'OpenRouterModel_{self._model_name}'
+
+  async def call(
+      self, content: AsyncIterable[content_api.ProcessorPartTypes]
+  ) -> AsyncIterable[content_api.ProcessorPartTypes]:
+    """Process content through OpenRouter API."""
+    messages = []
+    async for part in content:
+      if isinstance(part, content_api.ProcessorPart):
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:280]: You don't need that. You can safely change type annotation of `call` method to accept only `content_api.ProcessorPart`. The library laready has built-in conversion. When you use processors you can all the things convertible to ProcessorParts e.g. strings or PIL images. But when you write the processor itself, you get homogenous `ProcessorPart` objects. So implementation of your `call` method can be much simpler.

File: genai_processors/contrib/openrouter_model.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+  max_tokens: int | None
+  """Maximum number of tokens to generate."""
+
+  response_format: dict[str, Any] | None
+  """Response format specification (e.g., {"type": "json_object"})"""
+
+  stop: list[str] | str | None
+  """Stop sequences to end generation."""
+
+  stream: bool
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:108]: Is there a reason to not stream? I suggest to remove this option and always stream.

Non streaming (unary) option is usually there to simplify the API for a common use case, and especially if you deal with raw HTTP. But in Python we can provide better options. E.g. dealing with multimodal content forces us to use `Parts`, so your code has to deal with a list[Part] anyway. And iterable[Part] (which streaming provides) is not that much different.

File: genai_processors/contrib/openrouter_model.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        payload[key] = value
+
+    try:
+      # Make streaming request
+      async with self._client.stream(
+          'POST',
+          '/chat/completions',
+          json=payload,
+      ) as response:
+        response.raise_for_status()
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:309]: raise_for_status by design does not include the error returned by the server. this makes triaging the error harder. In Ollama integration we did this:
http://google3/third_party/py/genai_processors/core/ollama_model.py;l=242;rcl=780667442
I don't know how OpenRouter returns errors, but maybe a similar handling could be added?

File: genai_processors/contrib/openrouter_model.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+            # Handle finish reason
+            finish_reason = choice.get('finish_reason')
+            if finish_reason:
+              yield content_api.ProcessorPart(
+                  '',
+                  role='model',
+                  metadata={
+                      **self._build_metadata(parsed),
+                      'finish_reason': finish_reason,
+                      'generation_complete': True,
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:368]: We have END_OF_TURN part: https://github.com/google-gemini/genai-processors/blob/main/genai_processors/content_api.py#L587

Basically, use `'end_of_turn': True` instead of `'generation_complete': True,`

File: genai_processors/contrib/__init__.py [hunk 9]
```diff
@@ -12,4 +12,3 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
-"""Community contributed processors."""
```
Comment (kibergus) [inline genai_processors/contrib/__init__.py:15]: This line should remain.

File: genai_processors/contrib/README.md
Comment (kibergus) [inline genai_processors/contrib/README.md:1]: Having documentation is great. But the README.md is shared across all processors in contrib. Putting all the documentation in-to one file won't scale. Maybe put it in `openrouter_model.md`? And in `README.md` we can have links to such per-processor documents.

**Comment by kibergus** [comment]
Thanks for the pull request, it is a nice integration and it arrived amazingly fast. I'm still reading through it and polishing various parts, but it is getting late and I'll need to finish this tomorrow. I also want to try running it myself.

File: genai_processors/contrib/openrouter_model.py [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+import re
+from typing import Any, Literal
+
+from genai_processors import content_api
+from genai_processors import mime_types
+from genai_processors import processor
+from google.genai import types as genai_types
+import httpx
+from typing_extensions import TypedDict
+
```
Comment (nedalaltiti) [inline genai_processors/contrib/openrouter_model.py:63]: **Issue**: The current implementation uses generic `RuntimeError` for all API errors, making it difficult for users to handle specific error cases programmatically !!

Place these after all imports but before the _DEFAULT_BASE_URL

```python
class OpenRouterAPIError(Exception):
    """Base exception for OpenRouter API errors."""
    pass


class AuthenticationError(OpenRouterAPIError):
    """Raised when API authentication fails."""
    pass


class RateLimitError(OpenRouterAPIError):
    """Raised when API rate limit is exceeded."""
    pass
```
to enables proper retry strategies for rate limits, Better debugging and error reporting

File: genai_processors/contrib/openrouter_model.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+    
+    # Add OpenRouter-specific metadata
+    if 'id' in response_data:
+      metadata['request_id'] = response_data['id']
+    
+    if 'created' in response_data:
+      metadata['created'] = response_data['created']
+    
+    return metadata
+
```
Comment (nedalaltiti) [inline genai_processors/contrib/openrouter_model.py:405]: Issue: Error parsing logic is embedded in the exception handler, violating single responsibility principle and making it hard to test.

Extract error parsing to dedicated method by the following 


```python
def _parse_error_response(self, error_body: bytes) -> str:
    """Parse error details from API response.

    Args:
        error_body: Raw error response body from the API.

    Returns:
        Human-readable error message extracted from the response.
    """
    try:
        error_data = json.loads(error_body)
        # OpenRouter typically returns errors in {"error": {"message": "..."}} format
        return error_data.get('error', {}).get('message', error_body.decode('utf-8'))
    except (json.JSONDecodeError, UnicodeDecodeError):
        return error_body.decode('utf-8', errors='replace')

File: genai_processors/contrib/openrouter_model.py [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+                  '',
+                  role='model',
+                  metadata={
+                      **self._build_metadata(parsed),
+                      'finish_reason': finish_reason,
+                      'generation_complete': True,
+                  }
+              )
+
+    except httpx.HTTPStatusError as e:
```
Comment (nedalaltiti) [inline genai_processors/contrib/openrouter_model.py:372]: Issue: Current error handling loses important context and doesn't differentiate between error types.

```python
except httpx.HTTPStatusError as e:
    error_body = await e.response.aread()
    error_detail = self._parse_error_response(error_body)
    
    # Handle specific HTTP status codes with appropriate exceptions
    if e.response.status_code == 401:
        raise AuthenticationError(f"Invalid API key: {error_detail}") from e
    elif e.response.status_code == 429:
        retry_after = e.response.headers.get('Retry-After')
        raise RateLimitError(
            f"Rate limit exceeded. Retry after: {retry_after}s" if retry_after 
            else "Rate limit exceeded"
        ) from e
    else:
        raise OpenRouterAPIError(
            f"API request failed ({e.response.status_code}): {error_detail}"
        ) from e

**Comment by Seyamalam** [comment]
refactor: Address comprehensive maintainer feedback for OpenRouter integration

- Remove stream configuration option (always stream now)
- Use part.bytes instead of part.part.inline_data.data for image access
- Replace generation_complete with end_of_turn
- Use walrus operator for cleaner code (if content := delta.get('content'))
- Simplify call method signature and remove unnecessary type conversions
- Add custom exception hierarchy (OpenRouterAPIError, AuthenticationError, RateLimitError)
- Extract error parsing to dedicated _parse_error_response() method
- Handle specific HTTP status codes (401, 429) with appropriate exceptions
- Remove JSON error suppression to fail fast on malformed responses
- Add proper server error context and retry-after header handling
- Implement async context manager (__aenter__, __aexit__)
- Remove  __del__ method for resource cleanup
- Use response.aiter_lines() for cleaner streaming implementation
- Move README.md to processor-specific openrouter_model.md
- Add proper docstring to contrib/__init__.py
- Update all tests to match new error handling patterns
- Add async context manager test coverage
- Fix test expectations for new exception types
- All 19 tests passing with improved coverage

The integration now follows all library patterns, provides robust error
handling, and maintains clean resource management while addressing every
piece of maintainer feedback.
Thanks!

File: genai_processors/contrib/openrouter_model.py (hunk 10 above)
Comment (kibergus) [reply genai_processors/contrib/openrouter_model.py:62]: nedalaltiti, in general this is a good suggestion. But it also binds the user specifically to OpenRouterModel. But the errors themselves are not backend-specific. The problem is - once we submit this pull request, people will start using it and chinging exception type would become a breaking change. At the same time we need to balance the time/effort one has to put to submit things. We don't want to block on devising "proper" exception hierarchy.

For this reason I think re-raising httpx.HTTPStatusError is a good middle ground
https://github.com/google-gemini/genai-processors/blob/abe70d2c1afb4d5035965091f2052d436c397373/genai_processors/core/ollama_model.py#L246

File: genai_processors/contrib/openrouter_model.py [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+  top_a: float | None
+  """Alternative to top_p. Top-a sampling parameter."""
+
+  seed: int | None
+  """Random seed for deterministic outputs."""
+
+  max_tokens: int | None
+  """Maximum number of tokens to generate."""
+
+  response_format: dict[str, Any] | None
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.py:116]: Can this one be `genai_types.SchemaUnion` like in GenaiModel and OllamaModel?
https://github.com/google-gemini/genai-processors/blob/abe70d2c1afb4d5035965091f2052d436c397373/genai_processors/core/ollama_model.py#L67

And here is how we can get JSON schema out of it: https://github.com/google-gemini/genai-processors/blob/abe70d2c1afb4d5035965091f2052d436c397373/genai_processors/core/ollama_model.py#L198

It reuses some nice code from Gemini SDK that allows passing Python enums or dataclasses as schema.

File: genai_processors/contrib/openrouter_model.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    api_key="your-openrouter-api-key",
+    model_name="google/gemma-3-27b-it",  # Example model
+    generate_content_config={
+        'temperature': 0.7,
+        'max_tokens': 500,
+    }
+)
+
+# Use synchronously
+input_content = [content_api.ProcessorPart("What is AI?")]
```
Comment (kibergus) [inline genai_processors/contrib/openrouter_model.md:31]: Nit: no need to wrap with ProcessorPart: just `["What is AI?"]`

File: examples/openrouter_example.py
Comment (kibergus) [inline examples/openrouter_example.py:1]: We should add the copyright header to files in the repository, like other files have, in particular we explictly want to state Apache 2.0 license.

File: examples/openrouter_example.py [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+This example shows how to use the OpenRouter processor to access hundreds of
+AI models through a unified API. OpenRouter provides access to models from
+OpenAI, Anthropic, Google, Meta, and many other providers.
+
+## Setup
+
+To run this example, you'll need:
+
+1. An OpenRouter API key (get one at https://openrouter.ai/keys)
+2. Install the required dependencies
```
Comment (kibergus) [inline examples/openrouter_example.py:12]: Drop this line: there are no extra dependencies required. The core package actually already depends on httpx, I'll add it explicitly.

If there were dependencies to be installed, we should have said here what are they or how to figure out what to install.

**Review by kibergus — CHANGES_REQUESTED** [review]
Thanks. It is almost ready to be submitted. I've left some comments. Once they are resolved I'll merge your CL and probably will make some minor changes on top.

**Review by kibergus — CHANGES_REQUESTED** [review]
Thanks for your patience. I've pushed some amount of fixes/refactoring to the branch to address most of the comments.  @Seyamalam can you please take a look at the remaining comments and check that the code still works with actual servers.

After that I'll get this PR through our internal code review which is required to submit changes to our internal repository. I don't expect additional changes requested there, that would be mostly a formality, just would take a bit of time. Once I have that, I'll be able to merge the changes to github.

**Comment by Seyamalam** [comment]
Thank you for the detailed feedback! I've addressed all 5 issues you raised:

**Issue 1 - Redundant route parameter**: Removed the `route` field from `GenerateContentConfig` since it can only be 'fallback' and serves no purpose.

**Issue 2 - Incorrect docstring**: Fixed the `response_schema` docstring by removing the erroneous reference to `response_mime_type` which doesn't exist.

**Issue 3 - Function call handling**: Completely reworked the function call logic (lines 393-396) to:
- Use proper accumulation with `accumulated_function_call` buffer
- Add explicit error handling instead of silent `json.loads()` failures
- Only yield function calls when streaming completes
- Validate for incomplete/invalid function calls

**Issue 4 - Unnecessary metadata**: Removed `id` and `created` fields from `_build_metadata()` method, keeping only useful fields (`usage`, `model`).

**Issue 5 - Testing requirements**: Created `openrouter_model_test.py` with 7 comprehensive tests following the `ollama_model_test.py` pattern:
- `test_basic_inference`: Text streaming with metadata validation
- `test_headers_and_site_info`: Site URL/name parameters and complete header validation  
- `test_config_parameters`: Temperature, max_tokens, top_p validation
- `test_response_schema`: JSON schema format testing
- `test_function_calling`: Multi-turn conversation with function calls
- `test_image_input`: Base64 image handling
- `test_error_handling`: HTTP error scenarios

Also removed the outdated `openrouter_model.md` file as requested.

**Review by kibergus — APPROVED** [review]
